dm/pkg/checker/table_structure.go

```diff
@@ -131,30 +214,30 @@ func (c *TablesChecker) Check(ctx context.Context) *Result {
 		markCheckError(r, err)
 		return r
 	}
-	eg, checkCtx := errgroup.WithContext(ctx)
+
+	everyOptHandler, finalHandler := c.handleOpts(r)
+
+	pool := NewWorkerPoolWithContext[*checkItem, []*incompatibilityOption](
+		ctx, everyOptHandler,
+	)
+
 	for i := 0; i < concurrency; i++ {
-		eg.Go(func() error {
-			return c.startWorker(checkCtx)
-		})
+		worker := &tablesCheckerWorker{c: c}
+		worker.downstreamParser, err = dbutil.GetParserForDB(ctx, c.downstreamDB.DB)
+		if err != nil {
+			markCheckError(r, err)
+			return r
+		}
+		pool.Go(worker.handle)
 	}
-	// start consuming results before dispatching
-	// or the dispatching thread could be blocked when
-	// the output channel is full.
-	var optWg sync.WaitGroup
-	optWg.Add(1)
-	go func() {
-		defer optWg.Done()
-		c.handleOpts(ctx, r)
-	}()
-
-	dispatchTableItemWithDownstreamTable(checkCtx, c.tableMap, c.inCh)
-	if err := eg.Wait(); err != nil {
-		c.reMu.Lock()
+
+	dispatchTableItemWithDownstreamTable(c.tableMap, pool.PutJob)
+
+	if err := pool.Wait(); err != nil {
 		markCheckError(r, err)
-		c.reMu.Unlock()
+		return r
 	}
-	close(c.optCh)
-	optWg.Wait()
+	finalHandler()
```

Comment:
call it with `defer`?

Reply:
though its name is "finalXXX", it is a normal logic rather than some cleanup logic. It will append common message to the result where the message is calculated from all jobs, so if the result has been marked failed we can skip it.